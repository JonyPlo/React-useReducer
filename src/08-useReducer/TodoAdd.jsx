import { useForm } from '../hooks/useForm';
import PropTypes from 'prop-types';

export const TodoAdd = ({ onNewTodo }) => {
  const { description, onInputChange, onResetForm } = useForm({
    description: '',
  });

  const onFormSubmit = (event) => {
    event.preventDefault();

    const todoDescTrim = description.trim();
    if (todoDescTrim.length === 0)
      return alert('No se puede agregar una tarea vacia');

    const newTodo = {
      id: new Date().getTime(),
      description,
      done: false,
    };

    onNewTodo(newTodo);
    onResetForm();
  };

  return (
    <form onSubmit={onFormSubmit}>
      <input
        type='text'
        placeholder='Que hay que hacer?'
        className='form-control'
        name='description'
        value={description}
        onChange={onInputChange}
      />
      <button type='submit' className='btn btn-outline-primary mt-1'>
        Agregar
      </button>
    </form>
  );
};

TodoAdd.propTypes = {
  onNewTodo: PropTypes.func.isRequired,
};

TodoAdd.defaultProps = {
  onNewTodo: () => {
    console.log('No se admitio el prop al componente TodoAdd');
  },
};
