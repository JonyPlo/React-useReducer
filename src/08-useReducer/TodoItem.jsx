import PropTypes from 'prop-types';

export const TodoItem = ({
  id,
  description,
  done,
  onDeleteTodo,
  onToggleTodo,
}) => {
  return (
    <li
      key={id}
      id={id}
      className='list-group-item d-flex justify-content-between'
    >
      <span
        role='button'
        className={`align-self-center ${
          done ? 'text-decoration-line-through' : ''
        }`} aria-label='span'
        onClick={() => onToggleTodo(id)}
      >
        {description}
      </span>
      <button className='btn btn-danger' data-testid='deleteButton' onClick={() => onDeleteTodo(id)}>
        Borrar
      </button>
    </li>
  );
};

TodoItem.propTypes = {
  id: PropTypes.number.isRequired,
  description: PropTypes.string.isRequired,
  done: PropTypes.bool.isRequired,
};

TodoItem.defaultProps = {
  id: 12345,
  description: 'Esta es una descripcion por defecto',
  done: false,
};
