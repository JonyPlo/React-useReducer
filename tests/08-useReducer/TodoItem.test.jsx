import { fireEvent, render, screen } from '@testing-library/react';
import { TodoItem } from '../../src/08-useReducer/TodoItem';

describe('Pruebas en <TodoItem/>', () => {
  const todo = {
    id: 1,
    description: 'Piedra del Alma',
    done: false,
  };

  const onDeleteTodoMock = jest.fn();
  const onToggleTodoMock = jest.fn();

  beforeEach(() => jest.clearAllMocks());

  test('Debe de hacer match con el snapshot', () => {
    const { container } = render(<TodoItem />);
    expect(container).toMatchSnapshot();
  });

  test('Debe de mostrar el Todo pendiente de completar', () => {
    render(
      <TodoItem
        {...todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    );

    const liElement = screen.getByRole('listitem');

    expect(liElement.className).toBe(
      'list-group-item d-flex justify-content-between'
    );

    const spanElement = screen.getByLabelText('span');

    expect(spanElement.className).toContain('align-self-center');
    expect(spanElement.className).not.toContain('text-decoration-line-through');
  });

  test('Debe de mostrar el Todo completado', () => {
    todo.done = true;

    render(
      <TodoItem
        {...todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    );

    const spanElement = screen.getByLabelText('span');

    expect(spanElement.className).toContain('text-decoration-line-through');
  });

  test('El boton debe de llamar el deleteTodo', () => {
    render(
      <TodoItem
        {...todo}
        onToggleTodo={onToggleTodoMock}
        onDeleteTodo={onDeleteTodoMock}
      />
    );

    const deleteButton = screen.getByTestId('deleteButton');

    fireEvent.click(deleteButton);

    expect(onDeleteTodoMock).toHaveBeenCalledWith(todo.id);
  });
});
