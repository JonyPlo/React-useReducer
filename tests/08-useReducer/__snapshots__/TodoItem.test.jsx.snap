// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en <TodoItem/> Debe de hacer match con el snapshot 1`] = `
<div>
  <li
    class="list-group-item d-flex justify-content-between"
    id="12345"
  >
    <span
      aria-label="span"
      class="align-self-center "
      role="button"
    >
      Esta es una descripcion por defecto
    </span>
    <button
      class="btn btn-danger"
      data-testid="deleteButton"
    >
      Borrar
    </button>
  </li>
</div>
`;
